scipy/optimize/optimize.py

```diff
@@ -974,14 +992,40 @@ def check_grad(func, grad, x0, *args, **kwargs):
     >>> from scipy.optimize import check_grad
     >>> check_grad(func, grad, [1.5, -1.5])
     2.9802322387695312e-08
+    >>> rng = np.random.default_rng()
+    >>> check_grad(func, grad, [1.5, -1.5], 
+    ...             random_projection=True, seed=rng)
+    2.9802322387695312e-08
 
     """
+    x0 = np.asarray(x0)
     step = kwargs.pop('epsilon', _epsilon)
+    random_projection = kwargs.pop('random_projection', False)
```

Comment:
The use of `kwargs.pop()` is a legacy from Python 2, when Python did not support keyword only arguments other than  `**kwargs`. These should use keyword only arguments instead.

Reply:
It seems like we cannot mix `*args` and keyword only arguments. We cannot remove `*args` because they are needed by `func`, so possibly, passing `random_projection`, `seed` and `step` via `**kwargs` is the only possible way.

I tried the following signature but it doesn't seem to work,
```python
def check_grad(func, grad, x0, *args, *, step=_epsilon, random_projection=False, seed=None):
```
Error:
```bash
Traceback (most recent call last):
  File "/home/czgdp1807ssd/scipy_project/debug/check_grad.py", line 5, in <module>
    from scipy.optimize import check_grad
  File "/home/czgdp1807ssd/scipy_project/scipy/scipy/optimize/__init__.py", line 400, in <module>
    from .optimize import *
  File "/home/czgdp1807ssd/scipy_project/scipy/scipy/optimize/optimize.py", line 938
    def check_grad(func, grad, x0, *args, *, step=_epsilon, random_projection=False, seed=None):
                                          ^
SyntaxError: invalid syntax
```

Similarly, the following signature will not work (doesn't make sense but still I tried),
```python
def check_grad(func, grad, x0, *, step=_epsilon, random_projection=False, seed=None, *args):
```

Reply:
You can write:
`def check_grad(func, grad, x0, *args, step=_epsilon, random_projection=False, seed=None):`

Reply:
Okay. This will work. Thanks. I mis-interpreted that we are restricted to use keyword-only arguments (that after `*` thing). Making this extra stuff as optional arguments is fine too. Existing codes shouldn't face any problem with this.

Reply:
Anything after `*args` is also keyword only, whether or not it is optional :)